### ISSUE TITLE
Track longitudinal PyPI download stats for Dask and friends

### ISSUE THREAD

**Issue by ian-r-rose** [coiled/dask-community#13]
We currently have some older numbers for yearly and monthly download stats from PyPI. It would be nice to have some numbers and charts for those through time to track adoption and relative popularity.

A general plan for this could be:

1. Use the public [BigQuery](https://packaging.python.org/guides/analyzing-pypi-package-downloads/#google-cloud-bigquery) or similar to get some initial numbers and charts. Hand that off to marketing and evangelism.
2. Package the above into a notebook/gist of some sort so it can be easily re-run on demand
3. Possibly schedule the above re-running so we can regularly update the numbers and place the results in an easy-to-consume place.
4.

**Comment by ian-r-rose** [comment]
@jrbourbeau and @ncclementi might have some ideas about the best way forward here.

**Comment by ncclementi** [comment]
I'd have to take a deeper look but there might be a chance that we can read/run this with `dask-bigquery`. Looking at the dataset on bigquery, it is "bigquery partitioned".

What I'm not sure of is how we can read from the public data, for example in pure bigquery code: 

```python
from google.cloud import bigquery

project_name = "bigquery-public-data"
bq_dataset = "covid19_aha"  #this is a small dataset

client = bigquery.Client(project=project_name)

def gbq2df(sql):
    query = client.query(sql)
    results = query.result()
    return results.to_dataframe()
    
test = gbq2df("select * from covid19_aha.hospital_beds")
```
```
Forbidden: 403 POST https://bigquery.googleapis.com/bigquery/v2/projects/bigquery-public-data/jobs?prettyPrint=false: Access Denied: Project bigquery-public-data: User does not have bigquery.jobs.create permission in project bigquery-public-data.
```

I think this has to do with how we authenticate but I can't find information on how to solve the problem. If we can solve the problem on how to authenticate to be able to read public bigquery datasets then I think the rest should be straight forward.

**Comment by ian-r-rose** [comment]
Honestly, I think that `dask-bigquery` would be too powerful of a tool here. We really just want to create the bigquery `Client` and use it to execute a handful of queries that return aggregations. I don't think any postprocessing with a Dask cluster would be required.

But I know you have experience creating the google client objects, so I'd want to lean on that :)

**Comment by ncclementi** [comment]
We could use then just pure `bigquery` or `pandas-gbq` and I can write the code. But I'd need some help figuring out how to authenticate to be able to get the data from the public data sets. If we know someone that can help with that, that would be great.

**Comment by mrocklin** [comment]
When I've done this in the past I've just logged into the BigQuery web
query editor and run a query there.  If we wanted to automate this (which
isn't necessary for my personal short-term needs) then we would probably
use something like sqlalchemy, or some biguery API.

To be clear, short term I just want a timeseries of monthly downloads over
the past several years.  A quick google search yields
https://cloud.google.com/blog/topics/developers-practitioners/analyzing-python-package-downloads-bigquery

**Comment by ian-r-rose** [comment]
Notes from chatting with @scharlottej13 :

* We can do this for now by just using the google cloud console.
* We identified a basic query that does more or less what we want.
* We will get monthly download stats for (at least) the following projects
    * numpy
    * pandas
    * pyspark
    * dask
    * airflow
    * ray
    * vaex
    * modin
* The above can be downloaded into a CSV and added to a shared GDrive or similar.
* @scharlottej13 will then be able to create some nice charts/tables to share around

**Comment by scharlottej13** [comment]
Sticking the query here for reference. Results are on shared drive in OSS folder [here](https://drive.google.com/file/d/1KpdHz1frq54QeCdMNH8Z5KwjAVmT4un8/view?usp=sharing)
```
SELECT file.project as name,
COUNT(*) AS num_downloads,
DATE_TRUNC(DATE(timestamp), MONTH) AS `date`
FROM `bigquery-public-data.pypi.file_downloads`
WHERE
file.project IN('dask', 'ray', 'pandas', 'numpy', 'pyspark', 'modin', 'vaex', 'airflow')
-- query back to when Dask hit the scene
-- but doesn't seem to go back before 01/01/2016?
AND DATE(timestamp)
BETWEEN DATE('2015-01-01')
AND CURRENT_DATE()
GROUP BY `date`, file.project
ORDER BY file.project, `date` DESC
```

**Comment by scharlottej13** [comment]
monthly plots-- happy to make others!
![line_plt_log](https://user-images.githubusercontent.com/8620816/141599667-ef5dec05-0dcb-4471-a64e-febd9d5f9247.png)
![line_plt](https://user-images.githubusercontent.com/8620816/141599673-f405ca76-f98e-48f1-ace2-aab41ba1bdf9.png)

**Comment by mrocklin** [comment]
Woot.  Thanks Sarah.  These are great.  Can I also ask for just Dask,
linear y-scale?

**Comment by MrPowers** [comment]
@scharlottej13 - this is great.  Let me know if you'd like to collaborate on a blog post.  Would be a cool to give a little writeup for the PyData community. 🤓

**Comment by scharlottej13** [comment]
> Woot. Thanks Sarah. These are great. Can I also ask for just Dask, linear y-scale?
> […](#)
Of course! Here's just Dask. Not sure if granularity for the earlier years is interesting? If so we could try a broken y-axis or something.
![line_plt (1)](https://user-images.githubusercontent.com/8620816/141816823-2cae96c1-cb9d-42c6-bdb6-bc611609564d.png)

**Comment by scharlottej13** [comment]
> @scharlottej13 - this is great. Let me know if you'd like to collaborate on a blog post. Would be a cool to give a little writeup for the PyData community. 🤓

@MrPowers that'd be great! I'd be happy to collaborate on a blog post

**Comment by ian-r-rose** [comment]
I'd also be curious to see year-over-year (say Oct 20 vs Oct 21) changes in the number of monthly downloads for Dask, PySpark, and Ray: something like
|    |  Oct 2020 | Oct 2021 |
| -- | -------- | ----------- |
| Dask | x | y |
| PySpark | a | b |
| Ray | f | g |

**Comment by mrocklin** [comment]
Quick request, can I have the linear scale dask image, but with a transparent background?  (In general, transparent backgrounds are often welcome)

**Comment by scharlottej13** [comment]
> Quick request, can I have the linear scale dask image, but with a transparent background? (In general, transparent backgrounds are often welcome)

No problem! And good to know re transparent backgrounds

![line_plt](https://user-images.githubusercontent.com/8620816/142477858-ab736589-a7a5-40d8-87c3-aa49d4771b3e.png)

**Comment by scharlottej13** [comment]
> I'd also be curious to see year-over-year (say Oct 20 vs Oct 21) changes in the number of monthly downloads for Dask, PySpark, and Ray: something like
> 
> Oct 2020	Oct 2021
> Dask	x	y
> PySpark	a	b
> Ray	f	g

Sorry Ian, I somehow missed this yesterday! How's something like this for the numbers:

|    | 2018-10   | 2019-10   | 2020-10   | 2021-10    |
|---------|-----------|-----------|-----------|------------|
| dask    | 3,233,924 | 2,046,275 | 2,044,236 | 7,446,461  |
| pyspark | 500,282   | 2,466,990 | 9,679,279 | 14,396,151 |
| ray     | 23,315    | 311,345   | 344,673   | 596,116    |

And then the same thing, but percentage change between years, though this might be more meaningful for a larger time bin (maybe quarters are more meaningful?)

|  | Oct '18 to Oct '19   | Oct '19 to Oct '20   | Oct '20 to Oct '21  |
|---------|-----------|-----------|-----------|
| dask    | -37% |  -0.09%  | 264% |
| pyspark | 393%  | 292% |  49% |
| ray     | 1235% |  11% |  73%   |

**Comment by mrocklin** [comment]
I'm a little curious what went on with Dask last year.  is Is this what hitting mainstream is like?  Or is there just some more active bot out there having fun ?  :)

**Comment by scharlottej13** [comment]
I was going to ask you all the same! Are there other data sources I could look at to see if this is "real"? The numbers for pyspark also look a bit jumpy
